tf2onnx/rewriter/gemm_rewriter.py

```diff
@@ -0,0 +1,134 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT license.
+
+"""
+tf2onnx.rewrite - rewrite tensorflow subgraph to onnx gemm op
+"""
+
+from tf2onnx.graph_matcher import OpTypePattern, GraphMatcher
+
+# pylint: disable=missing-docstring
+
+def rewrite_gemm(g, ops):
+    if g.opset <= 6:
+        return ops
+
+    """
+    4 Candidate patterns are listed as follow, i.e. pattern0, pattern1, pattern2 and pattern 3
+    Where, A,B and C represent the three inputs, alpha and beta represent the two attributes.
+    """
+    # pattern0: alpha*A*B + beta*C
+    pattern0 = \
+        OpTypePattern('Add', name='add', inputs=[
+            OpTypePattern('Mul', name='mul1', inputs=[
+                OpTypePattern('Const', name='alpha'),
+                OpTypePattern('MatMul', name='matmul', inputs=[
+                    OpTypePattern('*', name='A'),
+                    OpTypePattern('*', name='B'),
+                ]),
+            ]),
+            OpTypePattern('Mul', name='mul2', inputs=[
+                OpTypePattern('Const', name='beta'),
+                OpTypePattern('*', name='C'),
+            ])
+        ])
+
+    # pattern1: alpha*A*B + C
+    pattern1 = \
+        OpTypePattern('Add', name='add', inputs=[
+            OpTypePattern('Mul', name='mul1', inputs=[
+                OpTypePattern('MatMul', name='matmul', inputs=[
+                    OpTypePattern('*', name='A'),
+                    OpTypePattern('*', name='B'),
+                ]),
+                OpTypePattern('Const', name='alpha')
+            ]),
+            OpTypePattern('*', name='C'),
+        ])
+
+    # pattern2: A*B + beta*C
+    pattern2 = \
+        OpTypePattern('Add', name='add', inputs=[
+            OpTypePattern('MatMul', name='matmul', inputs=[
+                OpTypePattern('*', name='A'),
+                OpTypePattern('*', name='B'),
+            ]),
+            OpTypePattern('Mul', name='mul2', inputs=[
+                OpTypePattern('*', name='C'),
+                OpTypePattern('Const', name='beta')
+            ]),
+        ])
+
+    # pattern3: A*B + C
+    pattern3 = \
+        OpTypePattern('Add', name='add', inputs=[
+            OpTypePattern('MatMul', name='matmul', inputs=[
+                OpTypePattern('*', name='A'),
+                OpTypePattern('*', name='B'),
+            ]),
+            OpTypePattern('*', name='C'),
+        ])
+
+    # pattern4: beta*C + alpha*A*B
```

Comment:
why still has this pattern?

Reply:
Thanks for your comment. I have updated the codes and delete the useless code comment.